namer/namer.cc

```diff
@@ -15,16 +18,1336 @@
+    NameKind kind;
+    FoundNameRef owner;
+    core::LocOffsets loc;
+    // The name of the modification.
+    core::NameRef name;
+    // For methods: The name of the method being modified.
+    core::NameRef methodName;
+};
+
+class FoundNames final {
+    // Contains references to items in _klasses, _methods, _staticFields, and _typeMembers.
+    // Used to determine the order in which symbols are defined in SymbolDefiner.
+    vector<FoundNameRef> _definitions;
+    // Contains references to classes in general. Separate from `FoundClass` because we sometimes need to define class
+    // Symbols for classes that are referenced from but not present in the given file.
+    vector<FoundClassRef> _klassRefs;
+    // Contains all classes defined in the file.
+    vector<FoundClass> _klasses;
+    // Contains all methods defined in the file.
+    vector<FoundMethod> _methods;
+    // Contains all static fields defined in the file.
+    vector<FoundStaticField> _staticFields;
+    // Contains all type members defined in the file.
+    vector<FoundTypeMember> _typeMembers;
+    // Contains all symbol references in the file.
+    vector<core::SymbolRef> _symbols;
+    // Contains all method and class modifiers (e.g. private/public/protected).
+    vector<Modifier> _modifiers;
+
+    FoundNameRef addDefinition(FoundNameRef ref) {
+        _definitions.push_back(ref);
+        return ref;
+    }
+
+public:
+    FoundNames() = default;
+    FoundNames(FoundNames &&names) = default;
+    FoundNames(const FoundNames &names) = delete;
+    ~FoundNames() = default;
+
+    FoundNameRef addClass(FoundClass &&klass) {
+        const u4 idx = _klasses.size();
+        _klasses.push_back(move(klass));
+        return addDefinition(FoundNameRef(NameKind::Class, idx));
+    }
+
+    FoundNameRef addClassRef(FoundClassRef &&klassRef) {
+        const u4 idx = _klassRefs.size();
+        _klassRefs.push_back(move(klassRef));
+        return FoundNameRef(NameKind::ClassRef, idx);
+    }
+
+    FoundNameRef addMethod(FoundMethod &&method) {
+        const u4 idx = _methods.size();
+        _methods.push_back(move(method));
+        return addDefinition(FoundNameRef(NameKind::Method, idx));
+    }
+
+    FoundNameRef addStaticField(FoundStaticField &&staticField) {
+        const u4 idx = _staticFields.size();
+        _staticFields.push_back(move(staticField));
+        return addDefinition(FoundNameRef(NameKind::StaticField, idx));
+    }
+
+    FoundNameRef addTypeMember(FoundTypeMember &&typeMember) {
+        const u4 idx = _typeMembers.size();
+        _typeMembers.push_back(move(typeMember));
+        return addDefinition(FoundNameRef(NameKind::TypeMember, idx));
+    }
+
+    FoundNameRef addSymbol(core::SymbolRef symbol) {
+        const u4 idx = _symbols.size();
+        _symbols.push_back(move(symbol));
+        return FoundNameRef(NameKind::Symbol, idx);
+    }
+
+    void addModifier(Modifier &&mod) {
+        _modifiers.push_back(move(mod));
```

Comment:
```suggestion
        _modifiers.emplace_back(move(mod));
```

You probably want `emplace_back` in every place you currently have `push_back` in this file.

Reply:
Is there a benefit to this?

Reply:
Saves a copy constructor invocation.

Reply:
ACK, I need to improve my intuition around this. I was under the impression that `push_back` was the more efficient of the two.